specs/WebResourceRequested-CustomScheme.md

```diff
@@ -0,0 +1,295 @@
+
+As a result, we introduce a new API to be able to register custom schemes and allow the end developer to answer these questions as a part of registration.
+
+# Conceptual pages (How To)
+
+WebResourceRequested event can also be raised for custom schemes. For this, the app has to register the custom schemes it would like to be able to issue resource requests for. With each registration, the app will specify whether the URIs with such schemes are considered secure context and it will also need to explicitly specify the origins that are allowed to make requests to these custom scheme URIs to ensure that untrusted origins cannot read trusted data from the app. The registrations are valid throughout the lifetime of the CoreWebView2Environment and browser process and any other CoreWebView2Environments that share the browser process must register exactly same schemes to be able to create a CoreWebView2Environment. The registered custom schemes also participate in [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) and [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP).
+
+For each custom scheme the developer wants to register they can:
+- Set it to be a secure context scheme to prevent insecure content warnings.
+- Provide a list of origins that can make requests to the custom scheme.
+
+# Examples
+
+## CustomSchemeRegistrations
+
+The following code sample registers 2 custom schemes on the CoreWebView2EnvironmentOptions:
+
+- Scheme 1 has https://*.example.com in its allowed origins list.
+- Scheme 2 has a null allowed origins list.
+
+It adds a WebResourceRequested handler for the requests to the custom schemes, which will serve the requests locally. Then it navigates the WebView2 to https://www.example.com and attempts to use XMLHttpRequest (XHR) with both schemes. Only XHR to scheme 1 will succeed because it's the one that has https://www.example.com in its allowed origin list.
+``` c#
+CoreWebView2EnvironmentOptions options = new CoreWebView2EnvironmentOptions();
+string[] allowedOriginsList = new string[1];
+allowedOriginsList[0] = "https://*.example.com";
+
+options.CustomSchemeRegistrations.Add(new CoreWebView2CustomSchemeRegistration("custom-scheme", true /* isSecure */, allowedOriginsList));
+options.CustomSchemeRegistrations.Add(new CoreWebView2CustomSchemeRegistration("custom-scheme-not-in-allowed-origins", true /* isSecure */, null));
+...
+string customScheme = "custom-scheme";
+webView.CoreWebView2.AddWebResourceRequestedFilter(
+        customScheme + ":*", All);
+webView.CoreWebView2.WebResourceRequested += delegate (
+            object sender, CoreWebView2WebResourceRequestedEventArgs e) {
+        CoreWebView2WebResourceResponse response;
+        CoreWebView2WebResourceRequest request = e.Request;
+        String uri = request.uri;
+        if (uri.StartsWith(customScheme + "://"))
+        {
+            string assetsFilePath = L"data/";
+            assetsFilePath += uri.Substring(customScheme.Length + 3);
```

Comment:
I assume that `../` are collapsed out as part of resolution. If not, then we have the danger of `custom-scheme://../../windows/system32/sam/passwords` being used to steal passwords.

Reply:
Please ensure path '..' and '.' segments are resolved as a part of path normalization and document that.